File: iOS/Kalorie/Core/Utils/FavouriteToggling.swift
//
//  FavouriteToggling.swift
//  Kalorie
//
//  Created by Josef Antoni on 12.08.2026.
//

import Foundation

enum FavouriteToggleError: Error {
    case missingItem
}

protocol FavouriteToggling: AnyObject {
    var isFavourite: Bool { get set }
    var isTogglingFavourite: Bool { get set }
    var alertItem: AlertItem? { get set }
}

extension FavouriteToggling {

    @MainActor
    func toggleFavourite(
        item: FoodItemDomain?,
        removalId: String,
        addFavouriteFood: any AddFavouriteFoodUseCaseProtocol,
        removeFavouriteFood: any RemoveFavouriteFoodUseCaseProtocol,
        onToggled: ((Bool) -> Void)? = nil
    ) async {
        guard !isTogglingFavourite else { return }
        isTogglingFavourite = true
        defer { isTogglingFavourite = false }

        let newValue = !isFavourite
        isFavourite = newValue
        do {
            if newValue {
                guard let item else { throw FavouriteToggleError.missingItem }
                try await addFavouriteFood(item)
            } else {
                try await removeFavouriteFood(id: removalId)
            }
            onToggled?(newValue)
        } catch {
            Log.error(error, category: Constants.LogCategory.favourites)
            isFavourite = !newValue
            alertItem = AlertItem(title: L10n.AddFood.errorFavouriteFailed)
        }
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/utils/FavouriteToggling.kt
package antoni.kalorie.core.utils

import antoni.kalorie.R
import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.usecases.AddFavouriteFoodUseCaseProtocol
import antoni.kalorie.core.usecases.RemoveFavouriteFoodUseCaseProtocol
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow

sealed class FavouriteToggleError : Exception() {
    data object MissingItem : FavouriteToggleError()
}

interface FavouriteToggling {
    val isFavourite: MutableStateFlow<Boolean>
    val isTogglingFavourite: MutableStateFlow<Boolean>
    val alertItem: MutableStateFlow<AlertItem?>

    // MARK: - Functions

    suspend fun toggleFavourite(
        item: FoodItemDomain?,
        removalId: String,
        addFavouriteFood: AddFavouriteFoodUseCaseProtocol,
        removeFavouriteFood: RemoveFavouriteFoodUseCaseProtocol,
        onToggled: ((Boolean) -> Unit)? = null,
    ) {
        if (isTogglingFavourite.value) return
        isTogglingFavourite.value = true
        val newValue = !isFavourite.value
        isFavourite.value = newValue
        try {
            if (newValue) {
                addFavouriteFood(item ?: throw FavouriteToggleError.MissingItem)
            } else {
                removeFavouriteFood(removalId)
            }
            onToggled?.invoke(newValue)
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.FAVOURITES)
            isFavourite.value = !newValue
            alertItem.value = AlertItem(titleRes = R.string.addFood_error_favouriteFailed)
        } finally {
            isTogglingFavourite.value = false
        }
    }
}
